### PR TITLE
Adds Plasma Guns MJP EDITION (Cannon/Rifle)

### PR DESCRIPTION
**Pull request by Pariah919**
## About The Pull Request
Adds two plasma guns, Cannon and Rifle.
tl;dr below, if you want the stats for nerds look at the drop down.

Plasma guns in general require their backpack to be able to be loaded, think an SG. I was originally going to add a belt but quickly realized that would be a bad idea, so consider this next part a plea to not let anyone add any other slot items beyond backpacks. **They are balanced with consideration  that they eat your back slot for operation.**

With that in consideration, consider this **testing** phase, heat is probably too high overall! But this matters little.

- Rifle - Has default cooling (2.5). Basically the bog standard.
- Normal - Fires a normal plasma shot.
- Overcharge - Fires a plasma shot that goes through mobs.
- Blast - Fires a blast of plasma with short range that applies incendiary, basically incendiary buckshot that applies melting.

Overall, Rifle aims for the bogstandard, it has higher damage than compareable acquantancies but due to heat it cannot fire that much.

- Cannon - Has poor cooling (1.5). Fire modes below.
- Normal - Machinegun style beat.
- Glob -  Fires fire globs, basically the same as a flamethrower.
- Charge - Shatters the target on contact for quite a while on hit, requires charging but is an incredibly fast projectile. 15 second shatter.

For acquisitions purposes, there are 5 of both in marine stock at roundstart, ammo is intended to be unlimited and if it is moved out of vendor then I'll just dump the plasma packs in the operational supply vendor.

**Plasma packs have 1000 charge.** **Chargepacks have 250.** TE powerpacks now only charge TE guns.

**Rifle down below.**

<details>

image

## Handling

- Aim Slowdown: 0.5 (Below average for a rifle, this is BR tier.
- Wield Delay: 0.4 SECONDS (VERY fast for a rifle, but this is offset by the fact you have to reconnect every time you get stunned.)
- Cool Amount: 2.5 (Average, should cool off in a reasonable amount of time.

## Normal

- Damage: 25
- Penetration: 15
- Sundering: 0.75
- Falloff: 1 (Double what most Rifles are since no falloff mult.)
- Power use on fire: 2.5 (400 shots if you use no other modes)
- Heat on fire: 3.3 (Totals up to 31~ shots in quick succession.)
- Fire Delay: 0.2 SECONDS
- Overall: This is a semi-limited req locked rifle stat profile, the damage of a T-12 and slightly higher pen for its class alongside sunder, this is a decent solo rifle that has to be taken in consideration with its fellow shot types.

## Piercing

- Damage: 30
- Penetration: 20
- Sundering: 2
- Falloff: 1
- Power use on fire: 12.5 (80 shots if you use no other modes)
- Heat on fire: 10 (Totals up to 10~ shots in quick succession.)
- Fire Delay: 0.25 SECONDS
- Overall: This is a very hard hitting heavy rifle, with decent sunder, higher than usual stats and the very very nice gimmick of being able to pierce through mobs, mainly will be used to shoot people or xenos being dragged off by allies.

## Blast

- Damage: 30
- Pellets: 4+1
- Pellet Damage: 25
- Penetration: 20
- Sundering: 2
- Falloff per pellet: 1x5 = 5
- Power use on fire: 50 (20 shots if you use no other modes)
- Heat on fire: 33 (Totals up to 10~ shots in quick succession.)
- Fire Delay: 2 SECONDS
- Melting per pellet: 1
- Overall: Trashed buckshot that deals melting stacks on every pellet hit. Enjoy.
</details>

**Cannon down below.**
<details>

image

## Handling

- Aim Slowdown: 1.2 (Same as GPMG.)
- Wield Delay: 1.5 SECONDS (Same as GPMG.)
- Cool Amount:  1.5 (Below average, better if you're using normal shot than the rifle I suppose.)

## Normal

- Damage: 20
- Penetration: 15
- Sundering: 0.75
- Falloff: 0.75
- Power use on fire: 2 (500 shots if you use no other modes)
- Heat on fire: 2 (Totals up to 50~ shots in quick succession.)
- Fire Delay: 0.15 SECONDS
- Overall: This is a semi-limited req locked MG stat profile being slightly worse than a GPMG, however you do get funny fire modes.
 
## Charge

- Damage: 120
- Penetration: 40
- Sundering: 10
- Falloff: 1
- Shell Speed: 4 (Higher than average.)
- Power use on fire: 100 (10 shots if you use no other modes)
- Heat on fire: 50 (Totals up to 2~ shots in quick succession.)
- Fire Delay: 1 SECONDS
- Shatter: 10 SECONDS
- Windup: 1 Second
- Overall: BREAK THEIR BONES THEIR ARMOR AND THEM.

## Glob

- Damage: 10
- Penetration: 100
- Sundering: N/A
- Falloff: 1
- Power use on fire: 100 (10 shots if you use no other modes)
- Heat on fire: 50 (Totals up to 2~ shots in quick succession.)
- Fire Delay: 2 SECONDS
- Flame Radius: 2
- Shell Speed: 2 (Below average.)
- Overall: Glob. With Flames. Its nice weed clear, by far the funniest part is now you can flame through *windows*. Do with this what you will.
</details>

## Why It's Good For The Game
Plasma guns are cool, and with the backpack gimmick in tow they allow for more interesting gun loadouts in a controlled case by case basis.
And no, I didn't make the previous PR! That was the French!
## Changelog
:cl:
add: Plasma guns, cannon and rifle, alongside the backpack used to power them, and the chargepacks used to power that.
add: Added a variable to powerpacks to only allow charging certain cell types, currently only applied to plasma chargepacks however.
balance: TE Powerpacks only charge TE lasgun cells.
/:cl:

### CODE REVIEW

**Review by Kuro020 — APPROVED** [review]
bless

**Comment by GuyGuysonF19** [comment]
Holy based

**Comment by Barnet2** [comment]
I am a MJP believer

**Comment by yyzsong** [comment]
2/10 evry stinky

**Comment by zzurge** [comment]
sezx

File: code/modules/projectiles/ammo_datums.dm [hunk 1]
```diff
@@ -3189,6 +3193,113 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
...
+
+/datum/ammo/energy/plasma/cannon_heavy
+	icon_state = "plasma_ball_big"
+	hud_state = "plasma_sphere"
+	damage = 120
+	penetration = 40
+	sundering = 10
+	damage_falloff = 1
+	shell_speed = 4
+	var/shatter_duration = 15 SECONDS
```
Comment (ivanmixo) [inline code/modules/projectiles/ammo_datums.dm:3270]: doc

Comment (ivanmixo) [reply code/modules/projectiles/ammo_datums.dm:3270]: also wowie that is extremely long

Comment (ivanmixo) [reply code/modules/projectiles/ammo_datums.dm:3270]: also this type is missing a name...

File: code/modules/projectiles/ammo_datums.dm [hunk 2]
```diff
@@ -3189,6 +3193,113 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
...
+
+/datum/ammo/energy/plasma/cannon_heavy/on_hit_mob(mob/M, obj/projectile/proj)
+	if(!isliving(M))
+		return
+	var/mob/living/living_victim = M
+	living_victim.apply_status_effect(STATUS_EFFECT_SHATTER, shatter_duration)
+
+
+/datum/ammo/energy/plasma/cannon_glob
+	name = "Plasma Glob"
```
Comment (ivanmixo) [inline code/modules/projectiles/ammo_datums.dm:3280]:
```suggestion
	name = "plasma glob"
```

**Review by ivanmixo — CHANGES_REQUESTED** [review]

**Review by ivanmixo — APPROVED** [review]

**Comment by 1kajowsmol** [comment]
got to use it on one round of crash so far; first suggestion i have so far is to remove the ability for them to mount underbarrel weapons, masterkey, pepperball, flamer, etc for definitely the cannon, and honestly maybe the rifle too.

**Comment by RagingLemons1** [comment]
fucked around with it in valhalla for ten minutes. thoughts:
fucking LMAO 
you added a gun that is an incin grenade launcher, a good light machinegun, and a railgun all in one that takes the same ammo, that also takes underbarrel slots. 
any problems with cooling can be mitigated by point blank buckshot to the face from a masterkey. 
time to kill on a primordial queen in valhalla was seven seconds on LMG mode, a single charge segment in LMG mode appears to be enough to kill or crit any xeno in the game, but was not tested on anything but hatchlings 
as it is these guns probably need a modes removed or SEVERE nerfs to actually be merged

**Comment by RagingLemons1** [comment]
it also one hits weeds with bayonets

**Comment by RagingLemons1** [comment]
the gun's incin mode does not appear to be able to kill walls

**Comment by yyzsong** [comment]
Mode switching doesn't put a message in the chat box. Heat generation feels too low, but that's just my taste.

**Comment by Pariah919** [comment]
```
you added a gun that is an incin grenade launcher, a good light machinegun, and a railgun all in one that takes the same ```

Well, you have no other guns at all, the overall stats could easily be adjusted.

```
that also takes underbarreled slots.```

Probably will remove cannon underbarreled, most people are saying its kinda ridiciolious and from a design perspective its inconsistent with the precedent of MGs.

```
 time to kill on a primordial queen in valhalla was seven seconds on LMG mode, a single charge segment in LMG mode appears to be enough to kill or crit any xeno in the game, but was not tested on anything but hatchlings
 ```
 
 Maturity isn't real? Is this your first time even bothering with clean roomtesting? You can get similar numbers with a GPMG, considering the only difference between both it and GPMG is 5 pen.
 
 
 ```
 the railgun mode appears to do significant sunder```
 
 It does 10. It has a one second windup and one second fire delay. Railgun for reference does HVAP 100 and 50 sunder, has no windup though this is closer to normal railgun shot which just does 150 damage and 20 sunder.
 This is a way slower projectile in general though.
 
 
 ```
  as it is these guns probably need a modes removed or SEVERE nerfs to actually be merged
  ```
  
  They cost you points, are limited, take up your back slot to able be able to be used and generally aren't that far removed from normal weapons stat wise? Beyond maybe things like piercing (Which might have too high damage in general still) and blast (which might be a meme still) alongside precedent breaking things like underbarreled on cannon, these aren't far stat wise from other req guns, they are 200 points after all.
  
  ```
  Mode switching doesn't put a message in the chat box. Heat generation feels too low, but that's just my taste.
  ```
  
  Okay to mode switching.
  Heat gen is what 1 magazine worth of stuff is for the weapon class, I think overheating giving you halved cooldown is way too lenient, will probably make it make your cool off time take longer. Generally speaking I think heat is on the high side considering how much easier normal guns are to wield and use.

**Comment by Texicos** [comment]
they all seem a bit weak considering the drawbacks of ammo scarcity and stuns disconnecting them for what is a little more penetration and worse handling overall on both. Swapping modes also keeps the fire delay so that doesn't really matter.

**Comment by Pariah919** [comment]
Pass 1
- Cannon no longer takes underbarreled guns, which fits with the precedent of other "machineguns". May be extended to rifle in the future.
- Adds Chargepacks, these recharge 25% of your local plasma guns, and fit everywhere a packet would.
- Codeside I added cell_type to make it so plasma chargepacks can only recharge plasma backpacks, so far I haven't applied this to volkite and laser pouches so they still recharge plasma bags. I'd rather not nerf either in an unrelated PR.

**Comment by yyzsong** [comment]
> I think overheating giving you halved cooldown is way too lenient, will probably make it make your cool off time take longer

Overheating simply preventing you from firing for a bit I feel doesn't do much due to how fast xenos move; by the time you overheat with most of these guns, they'll probably be gone. Most importantly, it's boring.

If it were up to me I'd give overheating a more severe consequence like burning your limbs. It feels thematically appropriate and it's not like it's hard to avoid overheating. Most importantly, it'd be way cooler.

**Comment by Pariah919** [comment]
> If it were up to me I'd give overheating a more severe consequence like burning your limbs. It feels thematically appropriate and it's not like it's hard to avoid overheating. Most importantly, it'd be way cooler.

That's dumb. No.

**Comment by yyzsong** [comment]
> That's dumb. No.

Since when has plasma guns burning you been a not cool thing dawg. In what fictional media

**Comment by blackdav123** [comment]
The cannon's fire lasts really long, I tested it a few times and a single click from the cannon can block off an area (no paths through the center 3x3) for 20-45 seconds. Similar weapons have nowhere near this much time, such as the plasma pistol having 14 seconds at most, the flamer having 20 at most, and the spray flamer having only 8. Considering the weapon's other benefits such as instant link to the backpack (unlike flamers long connect) this should definitely be reduced.

**Comment by Pariah919** [comment]
Pass 2
- Cannon can't take scopes.
- Puts Heat on glob to 50
- Glob has way worse fire stats.

**Comment by Wisemonster** [comment]
Does the blast mode on the rifle apply 1 melting per bullet? (So landing all of em will apply 5 stacks of melting) If so, that sounds really bad, considering that means marines get a max melting weapon at the start of each round, that can also remove pheros. I've also noticed that the overheat is so short, that using blast mode will result in about 0.2 seconds of being unable to fire it after it overheats, which is basically nothing.

https://github.com/tgstation/TerraGov-Marine-Corps/assets/87689371/4df8845a-5c93-42f8-b00f-57bc5da1189e

**Comment by XSlayer300** [comment]
Plasma guns.

T4 EORD random spawner slot.

Now.


image

**Comment by Arctrooper8** [comment]
> Plasma guns.
> 
> T4 EORD random spawner slot.
> 
> Now.
> 
> image

Holy shit lol, you actually told him. I'm sorry MJP but it's true you are infact a nerd.

**Comment by Pariah919** [comment]
Pass 3 (no more freebies)

- Glob Fire delay is 2 from 1.5
- Webbing fix (wew)
- Plasma Rifles/Cannons no longer spawn free, ideally these are made into HWR weapons in the future.

**Comment by XSlayer300** [comment]
where
image

>:(
